---
title: "Comparing variant calling precision and sensitivity between three vendors"
output: html_notebook
---

## Overview

Here we compare the sensitivity and precision of variant detection using three different vendors: Sentieon, GATK, and DRAGEN.
For each vendor, the same set of 17 samples (with the exception of Dragen - see below) was used for gVCF construction and joint-genotyping. The 17 samples included 16 samples from the lot 16-20 vantage run and a PCR-free bulk sample that has been subsampled to 450 million reads. 

```{r}

suppressPackageStartupMessages({
  library(tidyverse)
  library(grid)
  library(gridExtra)
})

```

```{r}
#Setup default themes
default_theme <- theme_light() + theme(axis.text.x = element_text(angle = -90), plot.title = element_text(hjust = 0.5))
```

## Internal gold-standard - Precision and Sensitivity of the three vendors

We load the precision and sensitivity calculations for the three vendors
```{r}
all_roc_filenames <- dir(path = "../scratch", pattern = "^NA12878", full.names = TRUE)
all_roc_internal_filenames <- all_roc_filenames[!grepl("-all_variants", all_roc_filenames)]
all_roc_internal_filenames <- all_roc_internal_filenames[!grepl("_vs_NA12878", all_roc_internal_filenames)]
all_roc_internal_filenames <- dir(all_roc_internal_filenames, pattern = "^snp_roc.tsv",full.names = TRUE)
internal_roc_df_list <- lapply(all_roc_internal_filenames, function(.filename) {
  temp_df <- read.table(.filename, stringsAsFactors = FALSE, comment.char = '', skip = 6, header = TRUE)
  sample_name <- sub("^.+_vs_(.+)-n16.+$", "\\1", .filename)
  sample_name <- sub("_combined", "\\1", sample_name)
  platform <- sub("^.+_", "", strsplit(.filename, "/")[[1]][3])
  temp_df['sample_name'] <- sample_name
  temp_df['platform'] <- platform
  return(temp_df)
})
internal_roc_df <- do.call(rbind, internal_roc_df_list)

group_by(internal_roc_df, platform, sample_name) %>% dplyr::filter(max(sensitivity) == sensitivity) %>% dplyr::filter(1:length(sensitivity) == length(sensitivity)) %>% dplyr::mutate(total_baseline_count = true_positives_baseline + false_negatives) %>% dplyr::ungroup()  %>% dplyr::arrange(platform)
#internal_comp_filenames <- all_roc_filenames[grepl("^NA12878.+-(n16_gatk)|(n16_dragen)|(n16_sentieon)", all_roc_filenames)]
#lapply()
#
#lapply()
#rtg_summaries_df <- read.table("../tracked_data/rtg_summary_all_samples.tsv", sep = "\t", header = TRUE) %>% dplyr::select(-X)
#dplyr::filter(rtg_summaries_df, sample_name == '4249-JW-10')
```

```{r fig.width = 14, fig.height = 10}
ggplot(internal_roc_df %>% dplyr::filter(sensitivity > 0.5, precision > 0.5), aes(1 - precision, sensitivity, color = platform)) + geom_point(size = 0.5) + facet_wrap( ~ sample_name, ncol = 5) + default_theme + labs(title = "ROC curves of 16 single-cell WGS datasets. All comparisons were made against a PCR-free bulk WGS sample.")
```

## External gold-standard - Precision and Sensitivity of the three vendors vs GIAB

```{r}
all_giab_roc_filenames <- dir(path = "../scratch", pattern = "^HG001", full.names = TRUE)
all_giab_roc_internal_filenames <- all_giab_roc_filenames[!grepl("-all_variants", all_giab_roc_filenames)]
all_giab_roc_internal_filenames <- all_giab_roc_internal_filenames[!grepl("_vs_NA12878", all_giab_roc_internal_filenames)]
all_giab_roc_internal_filenames <- dir(all_giab_roc_internal_filenames, pattern = "^snp_roc.tsv",full.names = TRUE)
internal_giab_roc_df_list <- lapply(all_giab_roc_internal_filenames, function(.filename) {
  temp_df <- read.table(.filename, stringsAsFactors = FALSE, comment.char = '', skip = 6, header = TRUE)
  sample_name <- sub("^.+_vs_(.+)-giab.+$", "\\1", .filename)
  sample_name <- sub("_combined", "\\1", sample_name)
  platform <- sub("^.+_", "", strsplit(.filename, "/")[[1]][3])
  temp_df['sample_name'] <- sample_name
  temp_df['platform'] <- platform
  return(temp_df)
})
internal_giab_roc_df <- do.call(rbind, internal_giab_roc_df_list)

group_by(internal_giab_roc_df, platform, sample_name) %>% dplyr::filter(max(sensitivity) == sensitivity) %>% dplyr::filter(1:length(sensitivity) == length(sensitivity)) %>% dplyr::mutate(total_baseline_count = true_positives_baseline + false_negatives) %>% dplyr::ungroup()  %>% dplyr::arrange(platform)
```

```{r fig.width = 14, fig.height = 10}
ggplot(internal_giab_roc_df %>% dplyr::filter(sensitivity > 0.5, precision > 0.5), aes(1 - precision, sensitivity, color = platform)) + geom_point(size = 0.5) + facet_wrap( ~ sample_name, ncol = 5) + default_theme + labs(title = "ROC curves of 16 single-cell WGS datasets. All comparisons were made against the GIAB NA12878 sample.")
```
